mpqp/core/instruction/gates/custom_gate.py

```diff
@@ -12,19 +17,49 @@
 
 @typechecked
 class CustomGate(Gate):
-    """Custom gates allow you to define your own gates.
+    """Custom gates allow you to define your own unitary gates.
 
     Args:
-        definition: The matrix (this is the only way supported to now) semantics of the gate.
+        definition: The UnitaryMatrix (only way supported for now) description of the gate.
         targets: The qubits on which the gate operates.
         label: The label of the gate. Defaults to None.
+
+    Raises:
+        ValueError: the target qubits must be contiguous and in order, and must match the size of the UnitaryMatrix
+
+    Example:
+        >>> u = UnitaryMatrix(np.array([[0,-1],[1,0]]))
+        >>> cg = CustomGate(u, [0])
+        >>> print(run(QCircuit([X(0), cg]), IBMDevice.AER_SIMULATOR))
+        Result: None, IBMDevice, AER_SIMULATOR
+         State vector: [-1, 0]
+         Probabilities: [1, 0]
+         Number of qubits: 1
+
+    Note:
+        For the moment, only ordered and contiguous target qubits are allowed when instantiating a CustomGate.
+
     """
 
     def __init__(
         self, definition: UnitaryMatrix, targets: list[int], label: Optional[str] = None
     ):
         self.matrix = definition.matrix
```

Comment:
We can indeed. To limit the influence on the rest of the code, I propose to define an attribute `self.unitary` and put the definition, and then define a property `self.matrix` that returns the matrix of the unitary.

This may however make pyright scream a bit if he can't understand that the type of self.matrix is Matrix, so will require to add some type cheking.

I let you confirm the idea before I implement.

Reply:
meh, not sure it's important enough to spend that energy. It's just that if one day we add other ways to define gates, it may be annoying to handle.

Maybe just a task on trello ?